### DISCUSSION TITLE
Issue Running Functional Tests

### DISCUSSION THREAD

**Discussion by champagne-cmd** [osm-search/Nominatim#2697] [Q&A] [answered]
Hello all, 

In attempting to address #2658, I have been attempting to run the functional tests currently in Nominatim, specifically the last three tests of the `addressing.feature` file. However, these three, as well as quite a few other tests, are failing (see attached file). It appears there is a link it is trying to access but can't. Any idea how to fix this?
[testrun3.txt](https://github.com/osm-search/Nominatim/files/8628373/testrun3.txt)

**Comment by lonvia** [comment, marked as answer]
Looks like you have introduced an error into the PHP part. The failure output of the test is somewhat misleading here.

The interesting part of the error message is this one:
```
      | osm | display_name            |
      | N1  | Bolder, Wonderway, Left |
      Assertion Failed: FAILED SUB-STEP: Then at least 1 result is returned
      Substep info: Traceback (most recent call last):
        File "/usr/lib/python3/dist-packages/behave/model.py", line 1329, in run
          match.run(runner.context)
        File "/usr/lib/python3/dist-packages/behave/matchers.py", line 98, in run
          self.func(context, *args, **kwargs)
        File "steps/steps_api_queries.py", line 201, in validate_result_number
          assert context.response.errorcode == 200
      AssertionError
```

It tells you that the data was successfully imported but sending a search query failed. The HTTP request returned with something other than 200. Unfortunately, it is not very helpful in telling you what exactly goes wrong.

The capture log that follows is not interesting for the error because it refers to the data import. And that was at this point successfully finished. You probably stumbled over that part:

```
      INFO:root:Using node id 2 for timestamp lookup
      CRITICAL:root:Failed to load URL: https://www.openstreetmap.org/api/0.6/node/2/1
      ERROR:root:Cannot determine date of database: HTTP Error 403: Forbidden
```
which is misleading, too, because it is not a critical error at all. It simply couldn't access the history information for the given node and therefore not find out about the date of the database. It's not really relevant for the tests and Nominatim happily works without the date.

So, how to debug the error in the meantime? I usually run the queries manually to see what the output is:

1. Run the failing test and keep the DB: `behave -DKEEP_TEST_DB db/import/addressing.feature:406`
2. Start Nominatm's development server with the test DB: `NOMINATIM_DATABASE_DSN=pgsql:dbname=test_nominatim nominatim serve`
3. Run nominatim-ui against `http://localhost:8088/` and search for `Bolder`.

I admit that it's all a bit annoying. Improving the BDD test behaviour on failures is on my list of things to do.

**Reply by champagne-cmd** [reply]
Thank you for the suggestions @lonvia , I really appreciate it! I've tried going through your steps, and I think I fixed some setup errors I had, but I'm still having trouble connecting the database and the UI. I am running step 2 above from a directory where I downloaded and set up a section of Germany (the one specific to #2649), and I am running nominatim-ui from its directory. However, there seems to be an issue with the ui fetching the data, or maybe just the setup of the ui:

 ```
chloe@chloe-VirtualBox:~/nominatim-ui-3.2.4/dist$ python3 -m http.server 8080
Serving HTTP on 0.0.0.0 port 8080 (http://0.0.0.0:8080/) ...
127.0.0.1 - - [05/May/2022 14:13:19] "GET /config.defaults.js HTTP/1.1" 304 -
127.0.0.1 - - [05/May/2022 14:13:19] "GET /theme/config.theme.js HTTP/1.1" 200 -
```

And on the server side: 

```
postgres@chloe-VirtualBox:/home/chloe/nominatim_example$ NOMINATIM_DATABASE_DSN=pgsql:dbname=test_nominatim nominatim serve
2022-05-05 14:19:22: Using project directory: /home/chloe/nominatim_example
[Thu May  5 14:19:22 2022] PHP 7.4.3 Development Server (http://127.0.0.1:8088) started
[Thu May  5 14:19:28 2022] 127.0.0.1:37670 Accepted
[Thu May  5 14:19:28 2022] 127.0.0.1:37670 [500]: GET /status.php?format=json
[Thu May  5 14:19:28 2022] 127.0.0.1:37670 Closing
```

An image of the webpage: 
![image](https://user-images.githubusercontent.com/54611898/167005306-8547132f-9fc6-4724-b1fb-679cf5c6c28b.png)
![image](https://user-images.githubusercontent.com/54611898/167011058-25ed0598-8686-4f36-84d2-c8864e282041.png)

I've tried switching the users running the commands, but all of the ones I've tried are superusers (postgres and chloe, I added chloe when setting up psql) and get errors thrown anyway. Can you tell from these errors if I'm missing some permissions somewhere?

**Reply by lonvia** [reply]
Looks like you are missing the ICU library for php. `sudo apt install php-intl` should do the trick, see [Transliterator in the php manual](https://www.php.net/manual/en/class.transliterator.php).
